baremetal/metal3machine_manager.go

```diff
@@ -1741,3 +1787,36 @@ func (m *MachineManager) getNodesWithLabel(ctx context.Context, nodeLabel string
 	}
 	return nodes, nodesCount, err
 }
+
+// getNodesWithoutLabel gets kubernetes nodes based on their Spec.providerID field.
+func (m *MachineManager) getNodesWithoutLabel(ctx context.Context, providerIDLegacy, providerIDNew string, providerIDonM3M *string, clientFactory ClientGetter) (int, error) {
+	corev1Remote, err := clientFactory(ctx, m.client, m.Cluster)
+	matchingNodesCount := 0
+	if err != nil {
+		return matchingNodesCount, errors.Wrap(err, "Error creating a remote client")
+	}
+
+	filter := metav1.ListOptions{}
+	nodes, err := corev1Remote.Nodes().List(ctx, filter)
+	if err != nil {
+		m.Log.Info(fmt.Sprintf("error while retrieving nodes: %v", err))
+		return matchingNodesCount, err
+	}
+	for _, node := range nodes.Items {
+		providerIDOnNode := node.Spec.ProviderID
+		if node.Spec.ProviderID == "" {
+			m.Log.Info("no providerID value found on node: ", node.GetName())
+		} else if providerIDOnNode == providerIDNew {
+			*providerIDonM3M = providerIDNew
+			matchingNodesCount = 1
+			return matchingNodesCount, nil
+		} else if providerIDOnNode == providerIDLegacy {
+			*providerIDonM3M = providerIDLegacy
+			matchingNodesCount = 1
+			return matchingNodesCount, nil
+		} else {
+			m.Log.Info("The node: ", node.GetName(), " does not match expected providerID: ", providerIDOnNode, "Considering other nodes ...")
+		}
```

Comment:
This logic is correct, but not fully resilient to the case where we would have a duplicated PID. This PR is supposed to fix this issue, but we should probably ensure that if two nodes match either the new or old providerID, we stop there and error out. This is needed for the transition phase

Reply:
Identifying the duplicates can be done. But, what the corrective measure in this case ? Try again ?